Optimize BBQ for runs of jump targets in `br.table`

danlliu opened the pull request:
#### d39f470ccbc0f2d146764be5e7a3f223cab49f30
<pre>
Optimize BBQ for runs of jump targets in `br.table`
<a href="https://bugs.webkit.org/show_bug.cgi?id=277686">https://bugs.webkit.org/show_bug.cgi?id=277686</a>
<a href="https://rdar.apple.com/133300128">rdar://133300128</a>

Reviewed by NOBODY (OOPS!).

Previously, BBQ generated an entire binary search for every possible case in a `br.table`,
even if we had long consecutive runs that could avoid many comparisons. This patch
optimizes `br.table` generation to reduce the tree depth when it is possible.

* Source/JavaScriptCore/jit/BinarySwitch.cpp:
(JSC::BinarySwitch::BinarySwitch):
(JSC::BinarySwitch::advance):
(JSC::BinarySwitch::buildCheckRuns):
(JSC::BinarySwitch::BranchCode::dump const):
* Source/JavaScriptCore/jit/BinarySwitch.h:
(JSC::BinarySwitch::BranchCode::BranchCode):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSwitch):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d39f470ccbc0f2d146764be5e7a3f223cab49f30

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61447 "check-webkit-style (failure)") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40796 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65390 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11984 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48474 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49620 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8320 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37914 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53215 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30463 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34575 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10444 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10897 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54539 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67119 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60686 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5382 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10512 "Found 1 new test failure: workers/worker-to-worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56992 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5405 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53180 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57209 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4441 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82452 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36600 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14398 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37683 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38777 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37428 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/707c8384a6f8ad3ac48917732772627070eb0166. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31786

kmiller68 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3692,7 +3692,7 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addSwitch(Value condition, const Vector
             bool isCodeEmitted = currentControlData().addExit(*this, target->targetLocations(), results);
             if (isCodeEmitted)
-                target->addBranch(m_jit.jump());
+        target->addBranch(m_jit.jump());
```
This whitespace change seems wrong.

kmiller68 commented on Source/JavaScriptCore/jit/WasmBinarySwitch.cpp:
```diff
@@ -0,0 +1,202 @@
+
+void WasmBinarySwitch::BranchCode::dump(PrintStream& out) const
+{
+    switch (kind) {
+    case NegativeToFallThrough:
+        out.print("NegativeToFallThrough");
+        break;
+    case NotLessThanToFallThrough:
+        out.print("NotLessThanToFallThrough");
+        break;
+    case LessThanToPush:
+        out.print("LessThanToPush");
+        break;
+    case Pop:
+        out.print("Pop");
+        break;
+    case ExecuteCase:
+        out.print("ExecuteCase");
+        break;
+    }
```
This could just be `out.print(kind);`

kmiller68 reviewed (changes requested):
I think we should integrate this into BinarySwitch rather than add a new class. That could be done by adding a new constructor if needed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8c5b442267502766e427a86b3f9ac09961f260a6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31786

danlliu committed:
Optimize BBQ for runs of jump targets in `br.table`
https://bugs.webkit.org/show_bug.cgi?id=277686
rdar://133300128

Reviewed by NOBODY (OOPS!).

Previously, BBQ generated an entire binary search for every possible case in a `br.table`,
even if we had long consecutive runs that could avoid many comparisons. This patch
optimizes `br.table` generation to reduce the tree depth when it is possible.

* Source/JavaScriptCore/jit/BinarySwitch.cpp:
(JSC::BinarySwitch::BinarySwitch):
(JSC::BinarySwitch::advance):
(JSC::BinarySwitch::buildCheckRuns):
(JSC::BinarySwitch::BranchCode::dump const):
* Source/JavaScriptCore/jit/BinarySwitch.h:
(JSC::BinarySwitch::BranchCode::BranchCode):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addSwitch):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d39f470ccbc0f2d146764be5e7a3f223cab49f30. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31786

hyjorc1 commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3710,18 +3710,24 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addSwitch(Value condition, const Vector
+        while (wasmSwitch.advance(m_jit)) {
+            unsigned value = wasmSwitch.caseValue();
+            unsigned index = wasmSwitch.caseIndex();
+            UNUSED_VARIABLE(value);
```
Just remove `value`?

kmiller68 replied:
Yeah, just get rid of `value`

kmiller68 commented on Source/JavaScriptCore/jit/BinarySwitch.cpp:
```diff
@@ -50,20 +51,31 @@ BinarySwitch::BinarySwitch(GPRReg value, std::span<const int64_t> cases, Type ty
-        m_cases.append(Case(cases[i], i));
-    
-    std::sort(m_cases.begin(), m_cases.end());
+    if (type == Int32 || type == IntPtr) {
```
Nit: I would add a `isCheckRuns()` and use that.

kmiller68 commented on Source/JavaScriptCore/jit/BinarySwitch.cpp:
```diff
@@ -50,20 +51,31 @@ BinarySwitch::BinarySwitch(GPRReg value, std::span<const int64_t> cases, Type ty
 #endif
+        build(0, false, m_cases.size());
+    } else {
+        m_cases.reserveInitialCapacity(cases.size());
```
`cases` are expected to be sorted right? Can we add an `ASSERT`?
